File: apps/api/app/schemas/attendance.py
import uuid
from datetime import date

from pydantic import BaseModel, ConfigDict, Field

STATUS_PATTERN = "^(present|absent|late|excused)$"


class AttendanceRecordRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    student_id: uuid.UUID
    attendance_date: date
    status: str
    marked_by_user_id: uuid.UUID | None = None
    remarks: str | None = None


class AttendanceMarkEntry(BaseModel):
    student_id: uuid.UUID
    status: str = Field(pattern=STATUS_PATTERN)
    remarks: str | None = None


class AttendanceBulkMarkRequest(BaseModel):
    attendance_date: date
    entries: list[AttendanceMarkEntry]


class PaginationMeta(BaseModel):
    page: int
    page_size: int
    total: int


class AttendanceListResponse(BaseModel):
    data: list[AttendanceRecordRead]
    meta: PaginationMeta


class AttendanceSummaryResponse(BaseModel):
    student_id: uuid.UUID
    total_days: int
    present_days: int
    absent_days: int
    late_days: int
    excused_days: int
    attendance_percentage: float | None


class AttendanceDefaulterRead(BaseModel):
    student_id: uuid.UUID
    student_name: str
    total_days: int
    present_days: int
    attendance_percentage: float


class AttendanceDefaultersResponse(BaseModel):
    threshold: float
    data: list[AttendanceDefaulterRead]

File: apps/web/src/types/attendance.ts
export type AttendanceStatus = "present" | "absent" | "late" | "excused";

export interface AttendanceRecord {
  id: string;
  student_id: string;
  attendance_date: string;
  status: AttendanceStatus;
  marked_by_user_id: string | null;
  remarks: string | null;
}

export interface AttendanceListResponse {
  data: AttendanceRecord[];
  meta: { page: number; page_size: number; total: number };
}

export interface AttendanceMarkEntry {
  student_id: string;
  status: AttendanceStatus;
  remarks?: string | null;
}

export interface AttendanceBulkMarkInput {
  attendance_date: string;
  entries: AttendanceMarkEntry[];
}

export interface AttendanceSummary {
  student_id: string;
  total_days: number;
  present_days: number;
  absent_days: number;
  late_days: number;
  excused_days: number;
  attendance_percentage: number | null;
}

export interface AttendanceDefaulter {
  student_id: string;
  student_name: string;
  total_days: number;
  present_days: number;
  attendance_percentage: number;
}

export interface AttendanceDefaultersResponse {
  threshold: number;
  data: AttendanceDefaulter[];
}
